core/shared/src/main/scala/zio/ZIOConcurrent.scala

```diff
@@ -0,0 +1,153 @@
+package zio
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.ConcurrentLinkedQueue
+import scala.annotation.tailrec
+import zio.Synchronizer._
+
+final class CountDownLatch(count: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(count, maxWaiters)
+
+  // TODO is it possible to evaluate once and reuse the Promise?
+  private[this] val awaitZero = sync.acquire(count)
+
+  def await: UIO[Unit] = awaitZero.asInstanceOf[UIO[Unit]]
+
+  // TODO stop at zero
+  def countDown(): UIO[Unit] = sync.release(1)
+}
+
+final class AsyncSemaphore(initialPermits: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(initialPermits, maxWaiters)
+
+  def apply[R, E, A](zio: ZIO[R, E, A]): ZIO[R, Either[E, Full], A] =
+    sync
+      .acquire(1)
+      .mapError(Right(_))
+      .flatMap { _ =>
+        zio
+          .mapError(Left(_))
+          .ensuring(sync.release(1))
+      }
+}
+
+final class ReentrantLock(maxWaiters: Option[Int]) {
+  @volatile private[this] var owner: Fiber.Id = null
+
+  private[this] val sem = new AsyncSemaphore(1, maxWaiters)
+
+  def apply[R, E <: Throwable, A](zio: ZIO[R, E, A]): ZIO[R, Either[E, Full], A] =
+    ZIO.effectSuspendTotalWith { (_, fiberId) =>
+      if (owner == fiberId) {
+        zio.mapError(Left(_))
+      } else {
+        sem(ZIO.effectSuspendTotal {
+          owner = fiberId
+          zio.ensuring(ZIO.effectTotal(owner = null))
+        })
+      }
+    }
+}
+
+final class Synchronizer(initialPermits: Int, maxWaiters: Int) {
+
+  // if >= 0, # of permits
+  // if < 0, # of waiters
+  private[this] val state = new AtomicInteger(initialPermits)
+
+  // TODO requires multiple allocations per waiter
+  private[this] val waiters = new ConcurrentLinkedQueue[(AtomicInteger, Promise[Full, Unit])]
+
+  def acquire(permits: Int): IO[Full, Unit] =
+    IO.effectSuspendTotalWith { (_, fiberId) =>
+      @tailrec def loop(): IO[Full, Unit] = {
+        val s = state.get()
+        if (s == -maxWaiters) {
+          // full
+          Synchronizer.full
+        } else if (s >= permits) {
+          // enough available permits
+          if (state.compareAndSet(s, s - permits))
+            Task.unit
+          else
+            loop()
+        } else {
+          // not enough permits, add waiter
+          val ns = if (s >= 0) -1 else s - 1
+          if (state.compareAndSet(s, ns)) {
+            val p   = Promise.unsafeMake[Full, Unit](fiberId)
+            val rem = if (s > 0) permits - s else permits
+            waiters.add((new AtomicInteger(rem), p))
+            p.await
+          } else {
+            loop()
+          }
+        }
+      }
+      loop()
+    }
+
+  def release(permits: Int): UIO[Unit] =
```

Comment:
Here too, if using the `Ref` with the `modify` trick, although you will still need a loop to dequeue as many waiters as possible.